tests/e2e/customsuites/multivolume.go

```diff
@@ -0,0 +1,306 @@
+package customsuites
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2evolume "k8s.io/kubernetes/test/e2e/framework/volume"
+	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+// s3CSIMultiVolumeTestSuite implements a test suite for multi-volume scenarios
+// with the S3 CSI driver, including sharing volumes between pods and mounting
+// multiple volumes in a single pod.
+type s3CSIMultiVolumeTestSuite struct {
+	tsInfo storageframework.TestSuiteInfo
+}
+
+// InitS3MultiVolumeTestSuite initializes and returns a test suite that validates
+// multi-volume functionality for the S3 CSI driver.
+//
+// This suite specifically tests:
+// - Multiple pods accessing the same S3 volume simultaneously
+// - A single pod accessing multiple S3 volumes concurrently
+// - Data persistence across pod recreations with the same volume
+//
+// The test suite verifies the core functionality needed for both stateless and
+// stateful workloads in Kubernetes when using S3 CSI volumes.
+func InitS3MultiVolumeTestSuite() storageframework.TestSuite {
+	return &s3CSIMultiVolumeTestSuite{
+		tsInfo: storageframework.TestSuiteInfo{
+			Name: "multivolume",
+			TestPatterns: []storageframework.TestPattern{
+				storageframework.DefaultFsPreprovisionedPV,
+			},
+		},
+	}
+}
+
+// GetTestSuiteInfo returns the test suite information including name and test patterns.
+func (t *s3CSIMultiVolumeTestSuite) GetTestSuiteInfo() storageframework.TestSuiteInfo {
+	return t.tsInfo
+}
+
+// SkipUnsupportedTests allows test suites to skip certain tests based on driver capabilities.
+// For S3 multi-volume scenarios, all tests should be supported, so this is a no-op.
+func (t *s3CSIMultiVolumeTestSuite) SkipUnsupportedTests(_ storageframework.TestDriver, _ storageframework.TestPattern) {
+}
+
+// DefineTests implements the test suite by defining all the test cases for multi-volume
+// scenarios. It creates the necessary volume resources, pods, and validation logic.
+func (t *s3CSIMultiVolumeTestSuite) DefineTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
```

Comment:
Should we add a test for concurrent access?

Reply:
Also, another good one as quite comment in real world is the RO, for read only

Reply:
For concurrency, we would like to do some smoke tests, rest should be tested in upstream mountpoint-s3

Reply:
Just to make sure I understand correctly, you meant concurrent access on a single volume right?
(This PR is for multi volume so confirming)

Reply:
Yes, multiple pods writing to a single volume :slightly_smiling_face: might not be relevant as you created a ticket for that.

The SEMANTICS md file also mentions that:
> This file access is optimized for applications that need high read throughput to large objects, potentially from many clients at once, and to write new objects sequentially from a single client at a time.

So maybe it's not supported on purpose, even from the same host (but the issue will also exist from multiple hosts I guess?), so +1 to have some tests

Reply:
Yes, its strictly for cheap storage tier
They do not care about performance and consistency (what you mentioned earlier, about 1st write and 2nd write).
The main use case is infinite storage, the main use case this project brings is we do not have re-provision volumes or need underlying storage to be the exact capacity (like block storage).